python/sensr_message_listener.py

```diff
@@ -39,78 +45,114 @@ def __init__(self,
+                        self._on_get_point_result(points)
+                    except asyncio.TimeoutError:
+                        pass
+                    except websockets.ConnectionClosedOK:
+                        pass
+                    except websockets.ConnectionClosedError:
+                        self._on_error("Closed by error.")
+
+    async def _main(self):
+        while self._state != MessageListener.State.STOPPED:
+            if self._state == MessageListener.State.STOP_REQUESTED:
+                await self.close_connection()
+                self._state = MessageListener.State.STOPPED
+            elif self._state == MessageListener.State.READY:
+                await self.close_connection()
+                self._state = MessageListener.State.RUNNING
+            else:
+                await asyncio.sleep(0.1)
+        
+        # Finalize
+        self._loop.stop()
+
+    async def close_connection(self):
+        if self._output_ws != None and self.is_output_message_listening():
+            await self._output_ws.close()
+            self._output_ws = None
+        if self._point_ws != None and self.is_point_result_listening():
             await self._point_ws.close()
-            # if object websocket is already closed, then call loop.stop()
-            # loop.stop should be called to exit `loop.run_forever`
-            if ((self._output_ws == None) or 
-                (self._output_ws != None and self._output_ws.closed == True)):
-                if (self._loop != None and self._loop.is_running()):
-                    self._loop.stop()
+            self._point_ws = None
 
     def connect(self):
-        print('Receiving SENSR output from {}...'.format(self._address))
-        try:
-            self._loop = asyncio.get_event_loop()
-        except asyncio.RuntimeError:
-            print('Fail to create event loop')
-            return False
-        self._is_running = True
-
-        if self.is_output_message_listening():
-            self._loop.create_task(self._output_stream())
-
-        if self.is_point_result_listening():
-            self._loop.create_task(self._point_stream())
-
-        self._loop.run_forever()
-        self._loop = None
-        return True
+        if self._state == MessageListener.State.STOPPED:
+            print('Receiving SENSR output from {}...'.format(self._address))
+            try:
+                self._loop = asyncio.get_event_loop()
+            except asyncio.RuntimeError:
+                print('Fail to create event loop')
+                return False
+            self._state = MessageListener.State.READY
+
+            if self.is_output_message_listening():
+                self._loop.create_task(self._output_stream())
+
+            if self.is_point_result_listening():
+                self._loop.create_task(self._point_stream())
+
+            self._loop.create_task(self._main())
+            self._loop.run_forever()
+            self._loop = None
```

Comment:
What does setting `self._loop` do?

Reply:
Do you mean `self._loop = None` line?
Then, nothing. Let me remove it.